### PR TITLE
SIGINT-896: Specific bridge download and caching

### PR DESCRIPTION
**Pull request by jahid1209**
- Added caching by parsing DOM so that development can move forward (will change the strategy after api endpoint is available.)

- Added functionality to download bridge from the url and version provided by the user.

- Added functionality to install bridge in the default location( Kishori's suggestion). If user provides the installation location, then bridge will be installed in the specified location.

- Added bridge parameter validation

- Addressed all of Kishori's review comment except this.

> If we download the zip to temp directory, we may not need to call delete.

This wasn't addressed because jvm doesn't erase the temp file even after the termination. Therefore, we need to call `delete` explicitly.

### CODE REVIEW

File: src/main/java/com/synopsys/integration/jenkins/scan/global/Utility.java [hunk 1]
```diff
@@ -13,8 +15,98 @@ public static FilePath createTempDir(String directoryName) throws IOException, I
...
+    }
+
+    public static File filePathToFile(FilePath filePath) {
+        String filePathString = filePath.getRemote();
+        File file = new File(filePathString);
+        return file;
+    }
+
+    public static void cleanupInputJson(String bridgeInstallationPath) {
+        File file = new File(bridgeInstallationPath, BridgeParams.BLACKDUCK_JSON_FILE_NAME);
```
Comment (zaman-akib) [inline src/main/java/com/synopsys/integration/jenkins/scan/global/Utility.java:106]: Have to keep in mind that later we may need to delete coverity and polaris json file as well

File: src/main/java/com/synopsys/integration/jenkins/scan/service/BlackDuckParametersService.java [hunk 2]
```diff
@@ -100,5 +106,4 @@ public Map<String, Object> createBlackDuckParametersMapFromJenkinsUI() {
```
Comment (zaman-akib) [inline src/main/java/com/synopsys/integration/jenkins/scan/service/BlackDuckParametersService.java:106]: Here we noticed a bug, if the global configuration is not done then `config.getBlackDuckUrl()` and `config.getBlackDuckCredentialsId()` produce null pointer exception. So we need to handle this. We may return the empty map if these values are empty or null.

File: src/main/java/com/synopsys/integration/jenkins/scan/SecurityScanner.java [hunk 3]
```diff
@@ -32,47 +32,72 @@ public SecurityScanner(TaskListener listener, Launcher launcher, FilePath worksp
...
+        if (blackDuckParametersService.performParameterValidation(blackDuckParameters)
+                && bridgeDownloadParametersService.performBridgeDownloadParameterValidation(bridgeDownloadParams)) {
+
+            FilePath bridgeInstallationPath = new FilePath(new File(bridgeDownloadParams.getBridgeInstallationPath()));
+            List<String> commandLineArgs = scannerArgumentService.getCommandLineArgs(bridgeInstallationPath, scanParams);
 
-        List<String> commandLineArgs = scannerArgumentService.getCommandLineArgs(workspace, scanParams);
+            BridgeDownloadManager bridgeDownloadManager = new BridgeDownloadManager();
+            boolean isBridgeDownloadRequired = bridgeDownloadManager.isSynopsysBridgeDownloadRequired(bridgeDownloadParams);
+//            BridgeDownloaderAndExecutor bridgeDownloaderAndExecutor = new BridgeDownloaderAndExecutor(listener, envVars);
```
Comment (zaman-akib) [inline src/main/java/com/synopsys/integration/jenkins/scan/SecurityScanner.java:53]: can be removed

File: src/main/java/com/synopsys/integration/jenkins/scan/global/Utility.java [hunk 4]
```diff
@@ -13,8 +15,98 @@ public static FilePath createTempDir(String directoryName) throws IOException, I
...
+                    deleteDirectory(file);
+                } else {
+                    file.delete();
+                }
+            }
+        }
+        directory.delete();
+    }
+
+    public static void copyRepository(String targetDirectory, String pluginWorkspaceDirectory) {
```
Comment (zaman-akib) [inline src/main/java/com/synopsys/integration/jenkins/scan/global/Utility.java:42]: We can alter the parameters place. 
`copyRepository(pluginWorkspaceDirectory, targetDirectory)` is easily understandable like `FileUtils.copyDirectory(workspaceDirectory, targetDir)`

File: src/main/java/com/synopsys/integration/jenkins/scan/bridge/BridgeDownloadManager.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import com.synopsys.integration.jenkins.scan.global.ApplicationConstants;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+public class BridgeDownloadManager {
+    private final String bridgeBinary = "synopsys-bridge";
+    private final String extensionsDirectory = "extensions";
+    private final String versionFile = "versions.txt";
```
Comment (zaman-akib) [inline src/main/java/com/synopsys/integration/jenkins/scan/bridge/BridgeDownloadManager.java:19]: should put a newline here

**Review by zaman-akib — CHANGES_REQUESTED** [review]

**Comment by jahid1209** [comment]
Thanks @zaman-akib for the feedback!

**Pull request merged**